### ISSUE TITLE
Question(investigate): isnt shutdownGracefully() sync() required?

### ISSUE THREAD

**Issue by oxplay2** [tlf30/monkey-netty#2]
Hello,

i noticed that for example Server have sync only for channel(tcp/udp) close:

```
            tcpConGroup.shutdownGracefully();
            tcpMsgGroup.shutdownGracefully();
            tcpFuture.channel().closeFuture().sync();
            udpConGroup.shutdownGracefully();
            udpMsgGroup.shutdownGracefully();
            ((UdpServerChannel) udpFuture.channel()).doClose();

```

isnt this needed to wait until groups will shutdown first? like:

```
            tcpConGroup.shutdownGracefully().sync();
            tcpMsgGroup.shutdownGracefully().sync();
            tcpFuture.channel().closeFuture().sync();
            udpConGroup.shutdownGracefully().sync();
            udpMsgGroup.shutdownGracefully().sync();
            ((UdpServerChannel) udpFuture.channel()).doClose();
```

Also i know its netty specific, but i try understand what exactly "cfg.setConnectTimeoutMillis" refer to.
it say:
> Sets the connect timeout of the channel in milliseconds. If the Channel does not support connect operation, this property is not used at all, and therefore will be ignored.

I have noticed that NettyClient will try reconnect if there are issues, but didnt noticed this timeout when server is off, so i guess its when server is on, but is unable to establish channel connection?

**Comment by tlf30** [comment]
Yes, the timeout does not appear to be working right now when the server goes offline, it is a problem I have been working on. But the reconnect does work, so I am still digging into it. 

As for the shutdowns, I do believe that the `tcpFuture.channel().closeFuture().sync();` will wait for the shutdowns. I could be wrong about that though. Have you seen where the tcp channel were still open for some time after the program exited?

**Comment by oxplay2** [comment]
Thanks,

about sync()

i never had this issue, but seen several stackoverflow.com issues like:

https://stackoverflow.com/questions/28032092/shutdown-netty-programmatically

where OP needed sync them.

The order is correct, as i seen in many places that channel need to be closed as last in order. 
But based on this link, should use sync().

Main question is, do work/main groups will finish their "works" before application will close. (maybe they are daemon threads, idk)

In any way, the people from this link had some issues, thats why i ask it here.

**Comment by tlf30** [comment]
Oh interesting. I will agree then, they should probably be added. Would you like to open a PR, or I can do it later tonight?

**Comment by oxplay2** [comment]
Problem:
- sync() for groups cause JME Window close take much longer time. (since it sync with main thread - cant do it in parallel)

Conclusion: If this issue would be real, then anyway would need probably cause executor thread to close it instead of Main JME thread. For now i think we should wait until we will be able replicate issue, so might close this issue, or leave it open.

**Comment by tlf30** [comment]
Sounds good. Lets leave this open then and if someone has a problem we can look back into it.
